src/vm.rs

```diff
@@ -683,18 +684,52 @@ impl<'a, S: Storage, C: PevmChain> Vm<'a, S, C> {
     }
 
     // Apply rewards (balance increments) to beneficiary accounts, etc.
-    fn apply_rewards(&self, write_set: &mut WriteSet, tx: &TxEnv, gas_used: U256) {
+    fn apply_rewards<DB: Database>(
+        &self,
+        write_set: &mut WriteSet,
+        tx: &TxEnv,
+        gas_used: U256,
+        #[allow(unused_variables)] evm_context: &EvmContext<DB>,
+    ) {
+        let mut gas_price = if let Some(priority_fee) = tx.gas_priority_fee {
+            std::cmp::min(tx.gas_price, priority_fee + self.block_env.basefee)
+        } else {
+            tx.gas_price
+        };
+        if self.spec_id.is_enabled_in(SpecId::LONDON) {
```

Comment:
We should add feature flags to all these spec checks? Not only for performance but to also avoid funny spec clashes between chains.

Reply:
That's something we can consider in future PRs.

I don't expect that removing an `if` statement improves the performance significantly.

If any benchmark proves the performance improvement, then yes we should do it. Otherwise it will just add noise to `Cargo.toml` and other `*.rs` files.

Reply:
No. Technically, there's no `LONDON` nor `SPURIOUS_DRAGON` in the context of OP and RISE. We'll even use a different Merkle tree for chain state with different state expiry rules, etc. 

Ignoring it will lead to bugs like skipping receipts root check for the first 4.3 millions blocks on new chains:
https://github.com/risechain/pevm/blob/3449a629ab3480ea2b4c57e710101f5dab751552/tests/common/runner.rs#L65-L74

These `>` checks are extremely slippery and not necessarily easy to notice when one mis-hits. An L2 can still proceed when the sequencer proposes a wrong state so we'll keep going when we mis-calculate a fee, receipt root, mis-clear a slot, etc., only for someone to challenge the wrong state transition and revert the chain plus fines.

So, `revm` still compiling here is a hack if anything. We (and users) must have full control of which chain applies which logic, including moving more spec abstractions towards our `Chain`. We're supposed to depend on Alloy types only and can support different underlying executors anyway.

A quick fix now is to use the `optimism` feature flag to either skip the check, or to use the correct spec (e.g., `BEDROCK` for OP.

Reply:
Not relevant to this PR, but this line:

```
if block.header.number >= 4370000
```

should be changed to:

```
if spec_id >= SpecId::BYZANTIUM
```

Reply:
Let's discuss offline tomorrow. I am not convinced.

> Technically, there's no LONDON nor SPURIOUS_DRAGON in the context of OP and RISE.

means `if self.spec_id.is_enabled_in(SpecId::LONDON) {` is an if statement where the condition is always true (because `SpecId::BEDROCK`, `SpecId::REGOLITH` are greater than `SpecId::LONDON`). So everything is fine!


> We (and users) must have full control of which chain applies which logic, including moving more spec abstractions towards our Chain.

Yes, if a chain has a specific logic, the logic should be put in `Chain`, not in the feature flag. Note that it's totally fine if someone compile with `--all-features` but then use `PevmEthereum::mainnet()`.

Reply:
Which we still need to make sure is only enabled for Ethereum. `BYZANTIUM` is a 6 so a new chain somehow starting from `0` and misputs it here is a risk we don't want.

Reply:
> means if self.spec_id.is_enabled_in(SpecId::LONDON) { is an if statement where the condition is always true (because SpecId::BEDROCK, SpecId::REGOLITH are greater than SpecId::LONDON). So everything is fine!

This literally means everything is not fine lol. I repeat this is a bad hack from `revm` because there's no linearity between chains' hardforks. What if RISE picks `SpecId::LATEST` but doesn't inherit the state removal of `SPURIOUS_DRAGON`? What if we launch with (some) `ECOTONE` specs of OP and some `PRAGUE` of Ethereum? Not to mention the risk of `revm` reordering things around and things still compile. 

Checking for `LONDON` when we have a feature flag to pick `BEDROCK` and saying "everything is fine because the condition is always true" is insane.

Reply:
Wait, the actual fix is to add to `Chain` things like `has_eip_161`, `has_eip_658`, etc. that take in a `SpecId`. The problem (of sloppy spec check) holds, but the solution of using feature flag was a tunnel vision from other discussions.

Reply:
> Wait, the actual fix is to add to Chain things like has_eip_161, has_eip_658, etc. that take in a SpecId. The problem (of sloppy spec check) holds, but the solution of using feature flag was a tunnel vision from other discussions.

Yes, that's what I mentioned in the last sentence of my last comment.

Feature flags (except OS, CPU related) are for optimization only, not for correctness. Anyone can compile with `--all-features` and expect the code to work.

Reply:
> add to Chain things like has_eip_161, has_eip_658

shouldn't this be done in another PR?

Reply:
The proper way is to have them added to `Chain`, replace the current checks, and merge it first. Then in this one we can focus on the implementation of OP, also as a clean reference of what a complete support looks like for new chains.

Reply:
See https://github.com/risechain/pevm/pull/347

Reply:
So, should the check be changed from:

```
if self.spec_id.is_enabled_in(SpecId::LONDON)
```

to 

```
if self.chain.is_eip_1559_enabled(self.spec_id)
```

?

----

Reference
https://eips.ethereum.org/EIPS/eip-1559

Reply:
Yes

Reply:
See https://github.com/risechain/pevm/pull/349

Reply:
Resolved by rebasing.